go/feast/featurestore.go

```diff
@@ -0,0 +1,907 @@
+	// 	return nil, err
+	// }
+
+	onlineFeatureResponse := &serving.GetOnlineFeaturesResponse{Metadata: &serving.GetOnlineFeaturesResponseMetadata{
+		FeatureNames: &serving.FeatureList{Val: make([]string, 0)},
+	},
+		Results: make([]*serving.GetOnlineFeaturesResponse_FeatureVector, numRows),
+	}
+	// Allocate memory for each GetOnlineFeaturesResponse_FeatureVector
+	for index = 0; index < numRows; index++ {
+		onlineFeatureResponse.Results[index] = &serving.GetOnlineFeaturesResponse_FeatureVector{Values: make([]*types.Value, 0),
+			Statuses:        make([]serving.FieldStatus, 0),
+			EventTimestamps: make([]*timestamppb.Timestamp, 0),
+		}
+	}
+	// Merge joinKeyValues into requestDataFeatures
+	for entityName, vals := range joinKeyValues {
+		requestDataFeatures[entityName] = vals
+	}
+
+	fs.populateResultRowsFromColumnar(onlineFeatureResponse, requestDataFeatures)
+
+	entitylessCase := false
+
+	for _, featureView := range featureViews {
+		if _, ok := featureView.entities[DUMMY_ENTITY_NAME]; ok {
+			entitylessCase = true
+			break
+		}
+	}
+
+	if entitylessCase {
+		dummyEntityColumn := &types.RepeatedValue{Val: make([]*types.Value, numRows)}
+		for index := 0; index < numRows; index++ {
+			dummyEntityColumn.Val[index] = &DUMMY_ENTITY
+		}
+		joinKeyValues[DUMMY_ENTITY_ID] = dummyEntityColumn
+	}
+
+	for table, requestedFeatures := range groupedRefs {
+		tableEntityValues, idxs, err := fs.getUniqueEntities(table, joinKeyValues, entityNameToJoinKeyMap)
+		if err != nil {
+			return nil, err
+		}
+		featureData, err := fs.readFromOnlineStore(tableEntityValues, requestedFeatures, table)
+		fs.populateResponseFromFeatureData(featureData,
+			idxs,
+			onlineFeatureResponse,
+			fullFeatureNames,
+			requestedFeatures,
+			table,
+		)
+	}
+	// TODO (Ly): ODFV, skip augmentResponseWithOnDemandTransforms
+	fs.dropUnneededColumns(onlineFeatureResponse, requestedResultRowNames)
+	return onlineFeatureResponse, nil
+}
+
+func (fs *FeatureStore) DestructOnlineStore() {
+	fs.onlineStore.Destruct()
+}
+
+func (fs *FeatureStore) parseKind(kind interface{}) (interface{}, error) {
+	if featureList, ok := kind.(*serving.GetOnlineFeaturesRequest_Features); ok {
+		return featureList.Features.GetVal(), nil
+	}
+	if featureServiceRequest, ok := kind.(*serving.GetOnlineFeaturesRequest_FeatureService); ok {
+		featureService, err := fs.registry.getFeatureService(fs.config.Project, featureServiceRequest.FeatureService)
+		if err != nil {
+			return nil, err
+		}
+		return featureService, nil
+	}
+	return nil, errors.New("Cannot parse kind from GetOnlineFeaturesRequest")
+}
+
+/*
+	This function returns all feature references from GetOnlineFeaturesRequest.
+	If a list of feature references is passed, return it.
+	Otherwise, FeatureService was passed, parse this feature service
+	to get a list of FeatureViewProjection and return feature references
+	from this list
+*/
+
+func (fs *FeatureStore) getFeatures(parsedKind interface{}, allowCache bool) ([]string, error) {
```

Comment:
It seems a bit strange that we are parsing the kind and still continue to pass around an `interface{}`. Is it possible for us to use a proper type for the kind?